### ISSUE TITLE
Target of assignment expression within list comprehension should belong to the containing scope

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#345]

**Comment by jakebailey** [comment]
Additional target scope / `:=` errors (feel free to split these out if needed).

```python
[x for x, y in (pairs2 := pairs) if x % 2 == 0]
[x for x, y in ([1, 2, 3, pairs2 := pairs]) if x % 2 == 0]
{x: y for x, y in (pairs2 := pairs) if x % 2 == 0}
{x for x, y in (pairs2 := pairs) if x % 2 == 0}
foo = (x for x, y in ([1, 2, 3, pairs2 := pairs]) if x % 2 == 0)

[[(j := j) for i in range(5)] for j in range(5)] # INVALID
[i := 0 for i, j in stuff]                       # INVALID
[i+1 for i in (i := stuff)]                      # INVALID

[False and (i := 0) for i, j in stuff]     # INVALID
[i for i, j in stuff if True or (j := 1)]  # INVALID

[i+1 for i in (j := stuff)]                    # INVALID
[i+1 for i in range(2) for j in (k := stuff)]  # INVALID
[i+1 for i in [j for j in (k := stuff)]]       # INVALID
[i+1 for i in (lambda: (j := stuff))()]        # INVALID

class Example:
    [(j := i) for i in range(5)]  # INVALID

[i for i in stuff if True or (j := 1) for j in range(10)]
```

Each line (except `class Example`) represents a "syntax error".

Missing things are:

- Assignment operator is not allowed in the "iterable" part of a comprehension.
  - `... for ... in (x := ...)` is an error at the `:=` expression, as `:=` is present in the `in`.
- Assignment operators on the left may not rebind iteration variable names.
  - `(x := ...) for x in ...` is an error at the `:=` expression, as `x` is already an iteration variable.
- Iteration variables may not rebind previously assignment-operator bound names.
  - `... for ... in ... if (x := ...) for x in ....` is an error at the `x` in `for x`, as `x` was already `:=` assigned.
- The class scope is not a valid target scope; the target must be either a function/lambda scope or the global scope.

**Comment by jakebailey** [comment]
```python
[[(j := j) for i in range(5)] for j in range(5)]
```

Is also unique in that currently it says "j is unbound" at the RHS `j` in the assignment expression, when it's bound by the outer comprension's `j`, but that may be a different binding bug.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next version. 
Here's the change:
https://github.com/microsoft/pyright/commit/d096acb5b2f95409bec1c09377ac0c0e5676605c

**Comment by erictraut** [comment]
This is now fixed in 1.1.3, which I just published.
